homeassistant/components/sensor/sigfox.py

```diff
@@ -0,0 +1,161 @@
+        if self.check_credentials():
+            device_types = self.get_device_types()
+            self._devices = self.get_devices(device_types)
+
+    def check_credentials(self):
+        """"Check API credentials are valid."""
+        url = urljoin(API_URL, 'devicetypes')
+        response = requests.get(url, auth=self._auth, timeout=10)
+        if response.status_code != 200:
+            if response.status_code == 401:
+                _LOGGER.error(
+                    "Invalid credentials for Sigfox API")
+            else:
+                _LOGGER.error(
+                    "Unable to login to Sigfox API, error code %s", str(
+                        response.status_code))
+            raise ValueError('Sigfox component not setup')
+        return True
+
+    def get_device_types(self):
+        """Get a list of device types."""
+        url = urljoin(API_URL, 'devicetypes')
+        response = requests.get(url, auth=self._auth, timeout=10)
+        device_types = []
+        for device in json.loads(response.text)['data']:
+            device_types.append(device['id'])
+        return device_types
+
+    def get_devices(self, device_types):
+        """Get the device_id of each device registered."""
+        devices = []
+        for unique_type in device_types:
+            location_url = 'devicetypes/{}/devices'.format(unique_type)
+            url = urljoin(API_URL, location_url)
+            response = requests.get(url, auth=self._auth, timeout=10)
+            devices_data = json.loads(response.text)['data']
+            for device in devices_data:
+                devices.append(device['id'])
+        return devices
+
+    @property
+    def auth(self):
+        """Return the API authentification."""
+        return self._auth
+
+    @property
+    def devices(self):
+        """Return the list of device_id."""
+        return self._devices
+
+
+class SigfoxDevice(Entity):
+    """Class for single sigfox device."""
+
+    def __init__(self, device_id, auth, name):
+        """Initialise the device object."""
+        self._device_id = device_id
+        self._auth = auth
+        self._message_data = {}
+        self._name = '{}_{}'.format(name, device_id)
+        self._state = None
+
+    def get_last_message(self):
+        """Return the last message from a device."""
+        device_url = 'devices/{}/messages?limit=1'.format(self._device_id)
+        url = urljoin(API_URL, device_url)
+        response = requests.get(url, auth=self._auth, timeout=10)
+        data = json.loads(response.text)['data'][0]
+        payload = bytes.fromhex(data['data']).decode('utf-8')
+        lat = data['rinfos'][0]['lat']
+        lng = data['rinfos'][0]['lng']
+        snr = data['snr']
+        epoch_time = data['time']
+        return {'lat': lat,
+                'lng': lng,
+                'payload': payload,
+                'snr': snr,
+                'time': epoch_to_datetime(epoch_time)}
```

Comment:
What kind of time point is this and how often does it change? Is it the time of the last message, ie it doesn't change on every update, or is it the time of the request, ie it changes every update?

We have a rule that says that a state is not allowed to change merely cause the time has changed. So we have to be careful when adding timestamps to the state machine.

Reply:
It is the time the message was published by the device